File: Scala/algorithmic_toolbox/src/main/scala-2.12/greedy/ChangingMoney.scala
package greedy

import scala.annotation.tailrec

object ChangingMoney {
  private val OrderedCoins: List[Int] = List(10, 5, 1)

  def calcMinimumNumberOfCoins(money: Int): Int = {
    @tailrec
    def loop(amount: Int, coins: List[Int], numberOfChanges: Int): Int = coins match {
      case Nil => numberOfChanges
      case largestCoin :: remainingCoins =>
        loop(amount % largestCoin, remainingCoins, numberOfChanges + amount / largestCoin )
      }
    loop(money, OrderedCoins, 0)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val amount: Int = reader.next().toInt
    val result: Int = calcMinimumNumberOfCoins(amount)
    println(result)
  }
}

File: OCaml/algorithmic_toolbox/greedy/changing_money.ml
let ordered_coins = [10; 5; 1]


let calc_minimum_number_of_coins amount =
    let rec inner money coins nr_changes = match coins with
        | [] -> nr_changes
        | coin :: remainingCoins -> inner (money mod coin) remainingCoins (nr_changes + money / coin)
    in inner amount ordered_coins 0


let () =
    let amount = read_int() in
    print_int (calc_minimum_number_of_coins amount);
    print_newline()
